File: day2.py
from dataclasses import dataclass

from util import load_input, bear_init


box_specs = load_input(2)


@bear_init
@dataclass
class Box:
    l: int
    w: int
    h: int

    @classmethod
    def from_str(cls, s: str) -> "Box":
        l, w, h = tuple(map(int, s.split("x")))
        return Box(l, w, h)

    @property
    def surface_area(self) -> int:
        return 2 * self.l * self.w + 2 * self.w * self.h + 2 * self.h * self.l

    @property
    def paper_area(self) -> int:
        """Total area of paper required.

        >>> Box.from_str("2x3x4").paper_area
        58

        >>> Box.from_str("1x1x10").paper_area
        43
        """
        return self.surface_area + min(
            self.l * self.w, self.w * self.h, self.h * self.l
        )

    @property
    def perimeters(self) -> tuple[int, int, int]:
        return (2 * (self.l + self.w), 2 * (self.w + self.h), 2 * (self.h + self.l))

    @property
    def ribbon_length(self) -> int:
        return min(self.perimeters) + self.l * self.w * self.h


boxes = [Box.from_str(spec) for spec in box_specs.splitlines()]

total_sqft = sum(box.paper_area for box in boxes)

print(f"Part 1: {total_sqft}")

total_ribbon_ft = sum(box.ribbon_length for box in boxes)

print(f"Part 2: {total_ribbon_ft}")

File: src/day2.jl
include("./util.jl")
using Test: @test

box_specs = util.load_input(2)

struct Box
    l::Int
    w::Int
    h::Int
end


function Box(s::AbstractString)::Box
    l, w, h = tuple(map(part -> parse(Int, part), split(s, 'x'))...)
    Box(l, w, h)
end

surface_area(b::Box)::Int = 2 * b.l * b.w + 2 * b.w * b.h + 2 * b.h * b.l

paper_area(b::Box)::Int = surface_area(b) + min(b.l * b.w, b.w * b.h, b.h * b.l)

@test paper_area(Box("2x3x4")) == 58
@test paper_area(Box("1x1x10")) == 43

perimeters(b::Box)::Tuple{Int,Int,Int} = (2 * (b.l + b.w), 2 * (b.w + b.h), 2 * (b.h + b.l))

ribbon_length(b::Box)::Int = min(perimeters(b)...) + b.l * b.w * b.h

boxes = map(Box, split(box_specs, '\n'))

total_sqft = sum(map(paper_area, boxes))
println("Part 1: $total_sqft")

total_ribbon_ft = sum(map(ribbon_length, boxes))
println("Part 2: $total_ribbon_ft")
